app/models/user.rb

```diff
@@ -1,16 +1,45 @@
 class User < ActiveRecord::Base
-  has_many :posts
-  validates :password, presence: true, length: {minimum: 8}
-  validates :username, presence: true, length: {minimum: 3}
-  validates :email, uniqueness: true, presence: true
+has_many :posts
+has_many :comments
+require 'bcrypt'
 =begin
-	User attributes
-
-	username: 	this is the username that the user goes by it can be used to log in. 
-	 			It showed be shown next to users posts and comments. S
-	 			Should be unique
-	password: 	personal password used to log in 
-	email: 		user's email addres it can be used to log in
+User attributes
+username: 	this is the username that the user goes by it can be used to log in. 
+	 		It showed be shown next to users posts and comments. S
+	 		Should be unique
+password: 	personal password used to log in 
+email: 		user's email addres it can be used to log in
 =end
 
+
+	EMAIL_REGEX =  %r{[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\z}i
+
+	validates :username, presence: true, uniqueness: true, length: { in: 3..20 }
+	validates :email, presence: true, uniqueness: true, format: EMAIL_REGEX 
+	
+	validates :password, confirmation: true, length: {in: 6..20}
+	validates :password_confirmation, presence: true
+	
+	has_secure_password	  
+
+	def self.get_username_by_id(id)
+		user = User.find_by_id(id)
+		if user
+			return user.username
+		else
+			return id
+		end
+	end
```

Comment:
This method shouldn't be necessary. Something like `User.find_by_id(whatever).try :username` will work. I don't see the value in this abstraction.